README.md
#Gateway Agent

The Gateway Agent (GA) is the element designed to make a bridge between the 
local Alljoyn bus and the Internet. It has a high level design that can be
found at the [Allseen Alliance page](
https://wiki.allseenalliance.org/gateway/gatewayagent).

This repository is organized with the following structure. There are 3 main
languages in use. One for computers, in C++, and the other two for mobiles,
specifically, iOS and Android (java).

The mobiles devices just have the Control App (using the name in the HLD),
named GatewayController in the source. On the other side, the C++ part, the
one to be working as the Gateway Agent, has the implementations for:

 - GatewayController -> Control App in the docs
 - GatewayConnector -> Connector App in the docs
 - GatewayMgmtApp -> Gateway Management App in the docs

As well as the code to create implementations of them, there are sample
implementations.

## Setting up the environment

To set up the environment, you just need to follow the Alljoyn environment
setup instructions. Taken directly from the [Allseen Alliance wiki](
https://wiki.allseenalliance.org/develop/contributing_source_code) you might
want to clone all the alljoyn tree:

```
mkdir aj-tree && cd aj-tree
repo init -u https://git.allseenalliance.org/gerrit/devtools/manifest.git
```

Once that is done, you can get into gateway/gwagent repository and run
scons.

ReleaseNotes.txt
GateWay agent 14.12 Release Notes
====================================

Platform Support
----------------
 1) Gateway Agent & sample connector
    - OpenWRT BB
    - Linux Ubuntu 14.04 LTS (64 bit)
 2) Controller
    - Android KitKat 4.4 (ARM7)
    - iOS 7.1 (32-bit)


Features added in Version 14.12
------------------------------

Issues Addressed in This Version 
--------------------------------
ASAGW-3	    Inconsistent naming convention in gateway public interface
ASAGW-4	    Inconsistent ways to return QStatus making it hard to use the gateway API
ASAGW-8	    Some mandatory fields are not initialised in AboutPropertyStoreImpl of GatewayAgent Manager App
ASAGW-11	XSD limits object path to 50 characters. This is too short. Should be longer or not limtied.
ASAGW-12    installPackage.sh shell script does not function on openwrt (TPLINK842ND) 
ASAGW-13	Add support for ALLJOYN_ECDHE_PSK
ASAGW-15	3rd party sample app stops after update of an ACL and GetMergedAcl method call
ASAGW-16    Getting error "Bad substitution" from sample app
ASAGW-17	Remoting Config interface fails due to ER_AUTH_FAIL at Exchange Guids.
ASAGW-20	Add about testing to connector sample app
ASAGW-21    Error thrown when ACL name is in UTF-8
ASAGW-24	IOS sample app project fails to compile due to wrong path in the project
ASAGW-26	Gateway Agent does not need onborading service
ASAGW-28    Policy entries not generated properly with some ACLs
ASAGW-29    Gateway Manager About App Name Misspelled
ASAGW-31    Documentation needed for GatewayConnector and GatewayController components
ASAGW-32    Output binaries and installation folders for Gateway need to conform to other AllJoyn projects
ASAGW-36    running postTweet.sh script with "sh -i" can hang on Ubuntu
ASAGW-41    Sample Connector App ipk not built properly for OpenWRT
ASAGW-42    Connector interface not registered as secured by Gateway Agent

A complete list of closed issues can be found at:
https://jira.allseenalliance.org/issues/?jql=project%20%3D%20ASAGW%20AND%20issuetype%20in%20%28Bug%29%20AND%20fixVersion%20%3D%20%2214.12%22%20AND%20status%20in%20%28Resolved%2CClosed%29%20ORDER%20BY%20key%20ASC


Known Issues
------------
ASAGW-18	New Gateway 3rd party apps are not registered without a restart of the GW agent.
ASAGW-25	Very small leak reported by Valgrind for Gateway Agent
ASAGW-40    GatewayMgmtApp should clean up "old" lost devices from policy db xml file sooner than a router restart

For a complete list of open issues:
https://jira.allseenalliance.org/issues/?jql=project%20%3D%20ASAGW%20AND%20status%20not%20in%20%28closed%2C%20resolved%29%20and%20type%20not%20in%20%28task%2C%20%22New%20Feature%22%29%20ORDER%20BY%20priority%20DESC%2C%20key%20ASC


Compatibility
-------------


Change history
--------------
14.12  - Initial Alliance Alliance release


cpp/GatewayMgmtApp/installPackage.sh
#!/bin/sh 

# Copyright (c) 2014, AllSeen Alliance. All rights reserved.
#
#    Permission to use, copy, modify, and/or distribute this software for any
#    purpose with or without fee is hereby granted, provided that the above
#    copyright notice and this permission notice appear in all copies.
#
#    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
#    WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
#    MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
#    ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
#    WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
#    ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
#    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
#

# sample script to handle installing of an application until package manager is available

# exit if the tar filename wasn't given
if [ -z "$1" ]; then
    echo "Usage: $(basename $0) [package.tar.gz]"
    exit 1
fi
tarFile=$1

# the base installation directory for alljoyn and the gateway agent
baseDir=/opt/alljoyn
tmpDir=
pkgInstallDir=
connectorId=

# whether install completed successfully
installCompleted=0
# whether we created a user
createdUser=0

# function to clean up files, etc. on exit
cleanup_on_exit() 
{
    # remove our temporary folder
    if [ -n "$tmpDir" ] && [ -d "$tmpDir" ]; then
        rm -fr $tmpDir
    fi

    if [ "$installCompleted" = 0 ]; then
        # remove the install directory (if we created it)
        if [ -n "$pkgInstallDir" ] && [ -d "$pkgInstallDir" ]; then
            rm -fr $pkgInstallDir
        fi
        # remove user (if we created it)
        if [ "$createdUser" != 0 ]; then
            echo userdel "$connectorId"
        fi
    fi
}

# add trap to clean up temp directory on exit
trap 'cleanup_on_exit' EXIT



echo "Trying to install: $tarFile"

# get a temp directory to untar the package
uuid=$(cat /proc/sys/kernel/random/uuid)
tmpDir=$baseDir/app-manager/$uuid

if [ -d "$tmpDir" ]; then
    echo "$tmpDir directory already exists. Should not have happened"; exit 6; 
fi

mkdir "$tmpDir" || exit 7

tar -xzf $tarFile -C $tmpDir

manifestFile=$tmpDir/Manifest.xml
if [ ! -f "$manifestFile" ]; then
    echo "Can not find $manifestFile file. Aborting"; exit 8; 
fi

appBinDir=$tmpDir/bin
if [ ! -d "$appBinDir" ]; then
    echo "Can not find $appBinDir directory. Aborting"; exit 9; 
fi

# get the connectorId from the Manifest.xml
connectorId=$(grep "<connectorId>" $manifestFile | sed -e "s/ *<connectorId>//" | sed -e "s/<\/connectorId *>//")

connectorAppDir=$baseDir/apps/$connectorId
if [ -d "$connectorAppDir" ]; then
    echo "Can not create directory for this app. It already exists: $connectorAppDir"; 
    exit 10; 
fi

pkgInstallDir=$connectorAppDir

mkdir -p $pkgInstallDir || exit 11
mkdir -p $pkgInstallDir/acls || exit 12
mkdir -p "$pkgInstallDir/bin" || exit 13
## TODO mkdir -p "$pkgInstallDir/etc || exit 14
mkdir -p "$pkgInstallDir/lib" || exit 15
## TODO mkdir -p "$pkgInstallDir/res" || exit 16
mkdir -p "$pkgInstallDir/store" || exit 17
## TODO mkdir -p "$pkgInstallDir/tmp" || exit 18

cp -rf $tmpDir/bin/* $pkgInstallDir/bin || exit 19
cp -rf $tmpDir/lib/* $pkgInstallDir/lib || exit 20
cp $tmpDir/Manifest.xml $pkgInstallDir/ || exit 21

# allow package to be installed for an existing user (for testing purposes)
id -u "$connectorId" &> /dev/null
if [ $? != 0 ]; then
    useradd $connectorId || exit 22
    createdUser=1
fi

chown -R "$connectorId" "$pkgInstallDir/store" || exit 23
chmod -R a+rx "$pkgInstallDir/bin" || exit 24
chmod -R a+rx "$pkgInstallDir/lib" || exit 25

installCompleted=1

echo "Successfully installed $tarFile with connectorId: $connectorId"; 
exit 0

